rust/kernel/spi.rs

```diff
@@ -0,0 +1,180 @@
+            this_module,
+            name,
+            probe,
+            remove,
+            shutdown,
+        ))?);
+
+        registration.as_mut().register()?;
+
+        Ok(registration)
+    }
+
+    // FIXME: Add documentation
+    pub fn register(self: Pin<&mut Self>) -> Result {
+        let mut spi_driver = bindings::spi_driver::default();
+        spi_driver.driver.name = self.name.as_ptr() as *const c_types::c_char;
+        spi_driver.probe = self.probe;
+        spi_driver.remove = self.remove;
+        spi_driver.shutdown = self.shutdown;
+
+        let this = unsafe { self.get_unchecked_mut() };
+        if this.registered {
+            return Err(Error::EINVAL);
+        }
+
+        this.spi_driver = Some(spi_driver);
+
+        let res = unsafe { bindings::__spi_register_driver(this.this_module.0, &mut spi_driver) };
+
+        match res {
+            0 => {
+                this.registered = true;
+                Ok(())
+            }
+            _ => Err(Error::from_kernel_errno(res)),
+        }
+    }
+}
+
+impl Drop for DriverRegistration {
+    fn drop(&mut self) {
+        unsafe { bindings::driver_unregister(&mut self.spi_driver.unwrap().driver) }
+        // FIXME: No unwrap? But it's safe?
+    }
+}
+
+// FIXME: Fix SAFETY documentation
+
+// SAFETY: The only method is `register()`, which requires a (pinned) mutable `Registration`, so it
+// is safe to pass `&Registration` to multiple threads because it offers no interior mutability.
+unsafe impl Sync for DriverRegistration {}
+
+// SAFETY: The only method is `register()`, which requires a (pinned) mutable `Registration`, so it
+// is safe to pass `&Registration` to multiple threads because it offers no interior mutability.
+unsafe impl Send for DriverRegistration {}
+
+type SpiMethod = unsafe extern "C" fn(*mut bindings::spi_device) -> c_types::c_int;
+type SpiMethodVoid = unsafe extern "C" fn(*mut bindings::spi_device) -> ();
+
+#[macro_export]
+macro_rules! spi_method {
+    (fn $method_name:ident (mut $device_name:ident : SpiDevice) -> Result $block:block) => {
+        unsafe extern "C" fn $method_name(dev: *mut kernel::bindings::spi_device) -> kernel::c_types::c_int {
+            use kernel::spi::SpiDevice;
+
+            fn inner(mut $device_name: SpiDevice) -> Result $block
+
+            match inner(SpiDevice::from_ptr(dev)) {
+                Ok(_) => 0,
+                Err(e) => e.to_kernel_errno(),
+            }
+        }
+    };
+    (fn $method_name:ident (mut $device_name:ident : SpiDevice) $block:block) => {
+        unsafe extern "C" fn $method_name(dev: *mut kernel::bindings::spi_device) {
+            use kernel::spi::SpiDevice;
+
+            fn inner(mut $device_name: SpiDevice) $block
+
+            inner(SpiDevice::from_ptr(dev))
+        }
+    };
+}
+
+pub struct Spi;
+
+impl Spi {
+    pub fn write_then_read(
+        dev: &mut SpiDevice,
+        tx_buf: &[u8],
+        n_tx: usize,
```

Comment:
Can't we get the length of a Rust slice using `.len()` ? If so, perhaps we can get rid of `n_tx` / `n_rx` ?

Reply:
That's true. We will rework it!

Reply:
Fixed by f0191b1